### PR TITLE
Fixed issue 14

### LINKED ISSUE #14: mkl-service should not set MKL_THREADING_LAYER environment variable
In `_mklinitmodule.c` the environment variable `MKL_THREADING_LAYER` is set. A library should never set any env vars.

This is resulting in issues in conda envs when `libgomp` is already loaded, see https://github.com/pytorch/pytorch/issues/37377.

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
Used `pthread_atfork` to restore environment variable `MKL_THREADING_LAYER` for the forked process.

@peterbell10 @rgommers

Closes #14 

Using the environment as specified in #14, and 

```python
# file iss14.py
def do_test(*args):
    import os
    print("in do_test   ", os.getenv("MKL_THREADING_LAYER", None))
    import torch; import numpy as np
    x = np.random.rand(100)
    x.dot(x)

if __name__ == '__main__':
    import multiprocessing as mp
    import os
    print("in __main__ before mkl   ", os.getenv("MKL_THREADING_LAYER", None))
    import mkl
    print("in __main__ after mkl   ", os.getenv("MKL_THREADING_LAYER", None))

    mp = mp.get_context('spawn')
    with mp.Pool(2) as p:
        p.map(do_test, range(2))
```

With this fix I am getting:

```
(debug_pytorch) [22:44:14 ansatnuc04 wd]$ python iss14.py
in __main__ before mkl    None
in __main__ after mkl    None
in do_test    None
in do_test    None
```

```
(debug_pytorch) [22:44:18 ansatnuc04 wd]$ MKL_THREADING_LAYER=GNU python iss14.py
in __main__ before mkl    GNU
in __main__ after mkl    GNU
in do_test    GNU
in do_test    GNU
```

### CODE REVIEW

File: mkl/_mklinitmodule.c [hunk 1]
```diff
@@ -29,30 +30,46 @@ static struct PyMethodDef methods[] = {
...
+}
+#endif
+
+static void _preload_threading_layer(void) {
+#if FORCE_PRELOADING
+#define SET_MTLAYER(L) do {						\
             VERBOSE("setting Intel(R) MKL to use " #L " OpenMP runtime\n");  \
             mkl_set_threading_layer(MKL_THREADING_##L);          \
             setenv("MKL_THREADING_LAYER", #L, 0);                \
+	    pthread_atfork(NULL, NULL, &restore_mtlayer);        \
```
Comment (anton-malakhov) [inline mkl/_mklinitmodule.c:59]: tabulation

**Review by anton-malakhov — APPROVED** [review]
LGTM

**Comment by rgommers** [comment]
Thanks @oleksandr-pavlyk!

**Comment by rgommers** [comment]
It would be good to know when this makes its way into conda `defaults` and `conda-forge`. For conda-forge I know where to look (https://github.com/conda-forge/mkl-service-feedstock). Not sure about `defaults`, except try `conda search mkl-service` sometime in the future. I assume this will be in the mkl-service `2.4.0` release?

**Comment by oleksandr-pavlyk** [comment]
Yes, I was thinking of release 2.4.0 today. I am looking through new functionality in MKL to add to service. 

[`mkl_set_num_stripes`](https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/support-functions/threading-control/mkl-set-num-stripes.html), `mkl_get_num_stripes`

Also adding support for more code in `mkl_enable_instructions` function.

**Comment by oleksandr-pavlyk** [comment]
@rgommers Fixed build break in https://github.com/conda-forge/mkl-service-feedstock/pull/17

**Comment by rgommers** [comment]
thanks!

**Pull request merged**